posts/views.py

```diff
@@ -16,12 +17,150 @@ def post_create(request):
     if request.method == 'POST':
         post_form = PostForm(request.POST, request.FILES)
         if post_form.is_valid():
-            post_form.save(commit=True)
+            post = post_form.save(commit=False)
+            post.owner = request.user
+            post.save()
+            for choice in post.choices:
+                if not Ingredients.objects.filter(name=choice):
+                    ingredient = Ingredients(name=choice)
+                    ingredient.save()
+                post.ingredients.add(Ingredients.objects.get(name=choice))
             return HttpResponseRedirect(reverse('posts:post_list'))
     return render(request, 'posts/post_create.html', context={'form': post_form})
 
 
-def post_list(request):
-    post_list = Post.objects.all()
-    return render(request, 'posts/post_list.html', context={'post_list': post_list})
+def post_list(request, pk=None):
+    if pk is None:
+        post_list = Post.objects.all()
+    else:
+        ingredient = Ingredients.objects.get(pk=pk)
+        post_list = ingredient.post_set.all()
+
+    ingredient_counts = {}
+    for ingredient in Ingredients.objects.all():
+        ingredient_counts[ingredient.name] = ingredient.post_set.count()
+    sorted_ingredient_counts = sorted(ingredient_counts.items(), key=operator.itemgetter(1))
+    sorted_ingredient_counts.reverse()
+    ingredients_list = []
+    for ingredient in sorted_ingredient_counts:
+        ingredients_list.append(Ingredients.objects.get(name=ingredient[0]))
+
+    for post in post_list:
+        num = 0
+        for rate in post.rate_set.all():
+            num += rate.rating
+        if post.rate_set.all():
```

Comment:
You should use `exists()` for this kind of controls.   `post.rate_set.all().exists()`